ledger/heavy/executor/backuper.go

```diff
@@ -0,0 +1,293 @@
+	}, nil
+}
+
+func move(ctx context.Context, what string, toDirectory string) error {
+	inslogger.FromContext(ctx).Debugf("backuper. move %s -> %s", what, toDirectory)
+	err := os.Rename(what, toDirectory)
+	if err != nil {
+		return errors.Wrapf(err, "can't move %s to %s", what, toDirectory)
+	}
+
+	return nil
+}
+
+// waitForBackup waits for file filePath appearance
+func waitForBackup(ctx context.Context, filePath string, numIterations uint) error {
+	inslogger.FromContext(ctx).Debug("waiting for ", filePath)
+	for i := uint(0); i < numIterations; i++ {
+		if err := isPathExists(filePath); err != nil {
+			if os.IsNotExist(err) {
+				inslogger.FromContext(ctx).Debugf("backup confirmation ( %s ) still doesn't exists. Sleep second.", filePath)
+				time.Sleep(time.Second)
+				continue
+			}
+			return errors.Wrap(err, "isPathExists return error")
+		}
+		return nil
+	}
+
+	return errors.New("no backup confirmation for pulse")
+}
+
+func writeBackupInfoFile(hash string, pulse insolar.PulseNumber, since uint64, upto uint64, to string) error {
+	bi := BackupInfo{SHA256: hash, Pulse: pulse, LastBackupedVersion: upto, Since: since}
+
+	rawInfo, err := json.MarshalIndent(bi, "", "    ")
+	if err != nil {
+		return errors.Wrap(err, "can't marshal backup info")
+	}
+
+	err = ioutil.WriteFile(to, rawInfo, 0600)
+	return errors.Wrapf(err, "can't write file %s", to)
+}
+
+func calculateFileHash(f *os.File) (string, error) {
+	hasher := sha256.New()
+	if _, err := io.Copy(hasher, f); err != nil {
+		return "", errors.Wrap(err, "io.Copy return error")
+	}
+	return fmt.Sprintf("%x", hasher.Sum(nil)), nil
+}
+
+// prepareBackup make incremental backup and write auxiliary file with meta info
+func (b *BackupMakerDefault) prepareBackup(dirHolder *tmpDirHolder, pulse insolar.PulseNumber) (uint64, error) {
+	currentBT, err := b.backuper.Backup(dirHolder.tmpFile, b.lastBackupedVersion)
+	if err != nil {
+		return 0, errors.Wrap(err, "Backup return error")
+	}
+
+	if err := dirHolder.reopenFile(); err != nil {
+		return 0, errors.Wrap(err, "reopenFile return error")
+	}
+
+	md5sum, err := calculateFileHash(dirHolder.tmpFile)
+	if err != nil {
+		return 0, errors.Wrap(err, "calculateFileHash return error")
+	}
+
+	metaInfoFile := filepath.Join(dirHolder.tmpDir, b.config.MetaInfoFile)
+	err = writeBackupInfoFile(md5sum, pulse, b.lastBackupedVersion, currentBT, metaInfoFile)
+	if err != nil {
+		return 0, errors.Wrap(err, "writeBackupInfoFile return error")
+	}
+
+	return currentBT, nil
+}
+
+type tmpDirHolder struct {
+	tmpDir  string
+	tmpFile *os.File
+}
+
+func (t *tmpDirHolder) release(ctx context.Context) {
+	err := t.tmpFile.Close()
+	if err != nil {
+		inslogger.FromContext(ctx).Fatal("can't close backup file: ", t.tmpFile, err)
+	}
+
+	err = os.RemoveAll(t.tmpDir)
+	if err != nil {
+		inslogger.FromContext(ctx).Fatal("can't remove backup file: ", t.tmpDir, err)
+	}
+}
+
+func (t *tmpDirHolder) reopenFile() error {
+	if err := t.tmpFile.Close(); err != nil {
```

Comment:
why reopen? why not just closing tmp file in regular code flow
unnecessary complexity.

Reply:
Where complexity is exactly?
I need to reopen file and save new fd in the same variable to close everything correctly later.
As user of tmpDirHolder I don't want to now about inner logic of opening and closing.